File: app/src/main/java/com/martinlaizg/geofind/data/repository/PlaceRepository.java
package com.martinlaizg.geofind.data.repository;

import android.app.Application;

import com.martinlaizg.geofind.data.access.api.service.PlayService;
import com.martinlaizg.geofind.data.access.database.AppDatabase;
import com.martinlaizg.geofind.data.access.database.dao.PlaceDAO;
import com.martinlaizg.geofind.data.access.database.dao.relations.PlacePlayDAO;
import com.martinlaizg.geofind.data.access.database.entities.Place;

import java.util.List;

public class PlaceRepository {

	private final PlaceDAO placeDAO;

	@SuppressWarnings("unused")
	PlaceRepository(Application application) {
		AppDatabase database = AppDatabase.getInstance(application);
		placeDAO = database.placeDAO();
		PlacePlayDAO placePlayDAO = database.playPlaceDAO();
		PlayService playService = PlayService.getInstance();
	}

	/**
	 * Insert a List of Place into the database
	 *
	 * @param places
	 * 		places to be inserted
	 */
	public void insert(List<Place> places) {
		for(Place p : places) insert(p);
	}

	/**
	 * Insert a single Place into the database
	 *
	 * @param place
	 * 		place to be inserted
	 */
	public void insert(Place place) {
		Place p = placeDAO.getPlace(place.getId());
		if(p == null) {
			placeDAO.insert(place);
		} else {
			placeDAO.update(place);
		}
	}

	/**
	 * Get the list of places, by tour id, from the database
	 *
	 * @param tour_id
	 * 		tour id of the places
	 * @return the list of places
	 */
	public List<Place> getTourPlaces(Integer tour_id) {
		return placeDAO.getTourPlaces(tour_id);

	}
}

File: app/src/main/java/com/martinlaizg/geofind/data/repository/PlaceRepository.kt
package com.martinlaizg.geofind.data.repository

import android.app.Application
import com.martinlaizg.geofind.data.access.database.AppDatabase
import com.martinlaizg.geofind.data.access.database.dao.PlaceDAO
import com.martinlaizg.geofind.data.access.database.entities.Place

class PlaceRepository private constructor(application: Application) {

	private val tag = PlaceRepository::class.simpleName
	private val placeDAO: PlaceDAO = AppDatabase.getDatabase(application).placeDAO()

	companion object {
		private var instance: PlaceRepository? = null
		fun getInstance(application: Application): PlaceRepository {
			return instance ?: synchronized(this) {
				val newInstance = PlaceRepository(application)
				instance = newInstance
				newInstance
			}
		}
	}

	/**
	 * Insert a List of Place into the database
	 *
	 * @param places
	 * places to be inserted
	 */
	fun insert(places: List<Place>) {
		for (p in places) insert(p)
	}

	/**
	 * Insert a single Place into the database
	 *
	 * @param place
	 * place to be inserted
	 */
	private fun insert(place: Place) {
		val p = placeDAO.getPlace(place.id)
		if (p == null) {
			placeDAO.insert(place)
		} else {
			placeDAO.update(place)
		}
	}

	/**
	 * Get the list of places, by tour id, from the database
	 *
	 * @param tourId
	 * tour id of the places
	 * @return the list of places
	 */
	fun getTourPlaces(tourId: Int?): MutableList<Place?>? {
		return placeDAO.getTourPlaces(tourId)
	}
}